library/index.c

```diff
@@ -530,10 +530,14 @@ int eblob_generate_sorted_index(struct eblob_backend *b, struct eblob_base_ctl *
 	struct eblob_map_fd src, dst;
 	int fd, err, len;
 	char *file, *dst_file;
+	int cache_is_empty = 0;
 
 	if (b == NULL || bctl == NULL)
 		return -EINVAL;
 
+	/* check that eblob cache is empty. If cache is empty we can skip iterating and flushing it */
+	cache_is_empty = eblob_cache_empty(b);
+
```

Comment:
> 1. There is a race between places where you check for empty cache and where this flag is used.  Besides the fact that it is a race, it leads to situations when key will not be deleted, overwrite will be lost and probably data is corrupted for example for partial writes.

You can't access/change/overwrite data from unsorted blob whose keys are not presented at the cache. And this situation is impossible.
> 2. Your change doesn't turn binlog assuming that since there are no keys in cache, no one can overwrite or delete data.

Now one can overwrite or delete data from usorted blob whose keys are not presented at the cache.
> Both assumptions may lead to data corruption. I can not accept this risk for the sake of initialization optimization.
>
> Btw, what are the benefits of that commit? How faster did blob start to load?
The benefit is a getting rid of useless iteration thru all blob's keys.